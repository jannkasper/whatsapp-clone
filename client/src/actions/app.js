
export const receiveAuthentication = payload => async dispatch => {
    dispatch({ type: "USER_AUTHENTICATION", payload });
};

export const openContactsNavigation = payload => async dispatch => {
    dispatch({ type: "OPEN_CONTACTS_NAVIGATION", payload });
};

export const closeContactsNavigation = payload => async dispatch => {
    dispatch({ type: "CLOSE_CONTACTS_NAVIGATION", payload });
};

export const setSelectedContact = payload => async dispatch => {
    dispatch({ type: "SET_SELECTED_CONTACT", payload });
};