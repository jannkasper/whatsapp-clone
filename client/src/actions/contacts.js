export const fetchContactsPending = () => ({ type: "FETCH_CONTACTS_PENDING" });


export const fetchContactsSuccess = payload => async dispatch => {
    dispatch({ type: "FETCH_CONTACTS_SUCCESS", payload });
};

export const fetchContactsError = payload => async dispatch => {
    dispatch({ type: "FETCH_CONTACTS_ERROR", payload });
};