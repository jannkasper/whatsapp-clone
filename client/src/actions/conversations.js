import { publicFetch } from "../util/fetcher";
import { prepare as prepareMessage } from "../util/message"
import { getSocket } from "../util/socket";

export const fetchConversationsPending = () => ({ type: "FETCH_CONVERSATIONS_PENDING" });

export const fetchConversationsSuccess = payload => async dispatch => {
    dispatch({ type: "FETCH_CONVERSATIONS_SUCCESS", payload });
};

export const fetchConversationsError = payload => async dispatch => {
    dispatch({ type: "FETCH_CONVERSATIONS_ERROR", payload });
};

export const selectConversation = payload => async dispatch => {
    dispatch({ type: "SELECT_CONVERSATION", payload });
};

export const createMessage = payload => async (dispatch, getState) => {
    const state = getState();
    const message = await prepareMessage(payload, state);

    dispatch({ type: "CREATE_MESSAGE", payload: { message: message.original }});

    getSocket().emit("MESSAGE", message.toSend);

    // publicFetch.post("message", message.toSend)
    //     .then(response => {
    //         if (!state.conversation.selectedConversation.sessionExtId) {
    //             dispatch({ type: "ENTER_SESSION_IDENTIFIER", payload: {sessionExtId: response.data.sessionExtId }});
    //         }
    //     });
}