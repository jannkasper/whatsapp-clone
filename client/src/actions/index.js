export * from "./app";
export * from "./contacts";
export * from "./messages";