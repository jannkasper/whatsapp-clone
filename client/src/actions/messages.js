import {publicFetch} from "../util/fetcher";

export const sendMessage = payload => async (dispatch, getState) => {
    const state = getState();

    const saveMessage = {
        type: payload.type,
        value: payload.value,
        status: 0,
        userExtId: state.user.externalIdentifier,
        created: Date.now()
    };
    dispatch({ type: "SEND_MESSAGE", payload: saveMessage });

    const messageData = {
        type: payload.type,
        value: payload.value,
        sessionExtId: state.conversation.selectedConversation.sessionExtId,
        userExtId: state.user.externalIdentifier,
        receiverExtId: state.conversation.selectedConversation.contactExtId,
        createdDate: Date.now()
    };
    const { data } = await publicFetch.post("message", messageData);

    if (!state.conversation.selectedConversation.sessionExtId) {
        dispatch({ type: "SET_CURRENT_SESSION_EXT_ID", payload: data.sessionExtId });
    }

}