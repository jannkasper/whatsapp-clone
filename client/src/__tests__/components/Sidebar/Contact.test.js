import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { format } from 'date-fns';
import configureStore from '../../../store';
import Contact  from '../../../components/Sidebar/contact/index';
import * as actions from '../../../actions';

const store = configureStore();

const params = {
    externalIdentifier: '1',
    name: 'nameTest',
    status: 'statusTest',
    lastMessage: {
        type: 'typeTest',
        value: 'valueTest',
        created: new Date(2020,10,10)
    },
    isSelected: false
}

jest.mock('../../../actions');

describe('Contact component', () => {
    it('should display', () => {
        const {asFragment} = render(
            <Provider store={store}>
                <Contact {...params} />
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    })

    it('can select contact', async () => {
        const setSelectedContactMock = jest.fn((contactExtId) => ({ type: '', contactExtId: contactExtId }));
        const selectConversationMock = jest.fn((contactExtId) => ({ type: '', contactExtId: contactExtId }));
        const closeContactsNavigationMock = jest.fn(() => ({type: ''}));

        jest.spyOn(actions, "setSelectedContact").mockImplementation(setSelectedContactMock);
        jest.spyOn(actions, "selectConversation").mockImplementation(selectConversationMock);
        jest.spyOn(actions, "closeContactsNavigation").mockImplementation(closeContactsNavigationMock);

        render(
            <Provider store={store}>
                <Contact {...params} />
            </Provider>,
        );

        const div = screen.getByRole('selectContact');

        // Validate
        fireEvent.click(div);

        await waitFor(() =>  {
            expect(setSelectedContactMock).toHaveBeenLastCalledWith({ contactExtId: '1' });
            expect(selectConversationMock).toHaveBeenLastCalledWith({ contactExtId: '1' });
            expect(closeContactsNavigationMock).toHaveBeenCalled();
        });
    });

    it('can determine date', async () => {
        render(
            <Provider store={store}>
                <Contact {...params} />
            </Provider>,
        );

        expect(screen.getByText(format(params.lastMessage.created, "dd/MM/yyyy"))).toBeDefined();

        // Update created date
        params.lastMessage.created.setDate(params.lastMessage.created.getDate() - 1);

        render(
            <Provider store={store}>
                <Contact {...params}/>
            </Provider>,
        );

        expect(screen.getByText(format(params.lastMessage.created, "dd/MM/yyyy"))).toBeDefined();
    });
});
