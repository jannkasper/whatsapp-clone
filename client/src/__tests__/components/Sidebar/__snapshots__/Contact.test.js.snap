// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact component should display 1`] = `
<DocumentFragment>
  <div
    class="contactContainer null"
    role="selectContact"
  >
    <div
      class="contactImage"
    >
      <img
        alt="profileImage"
        src="avatar.svg"
      />
    </div>
    <div
      class="contactDetails"
    >
      <div
        class="contactName"
      >
        <span>
          nameTest
        </span>
        <div>
          10/11/2020
        </div>
      </div>
      <div
        class="contactLastMessage"
      >
        <div
          class="contactLastMessage_inner"
        >
          <span
            class="span"
          >
            valueTest
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
